refactor(state): improve error propagation for CommitCheckpointVerifiedBlock

syszery opened the pull request:
## Motivation

This PR addresses #9730 and focuses on the `CommitCheckpointVerifiedBlock` path.


## Solution

<!-- Describe the changes in the PR. -->

* Refactors the errors in `CommitCheckpointVerifiedBlock` following the example from #9848 and #9923.
* Introduces the new error enum `CommitCheckpointVerifiedError` in `error.rs`.
* Implements `MappedRequest` for `CommitCheckpointVerifiedBlock`.
* Adds documentation for the new error type and its usage across relevant modules.


### Follow-up Work

* **Should `CloneError` be replaced?**
  `CloneError` was introduced due to [`commit_finalized_direct`](https://github.com/ZcashFoundation/zebra/blob/d1ce5e59b40ac09ca951b9d2c1a71f8037f11594/zebra-state/src/service/finalized_state.rs#L327) returning a `BoxError`. Replacing it will require additional changes to the error like implementing `From<zebra_chain::parallel::tree::NoteCommitmentTreeError>`. 
* **Should `CommitCheckpointVerifiedError` wrap `QueueAndCommitError`?**
  There is some overlap between these error types, particularly with regard to the error messages. Adopting `QueueAndCommitError` would simplify error handling, but this would require modifying some of the existing error messages to match.

Happy to adjust based on feedback.


### PR Checklist

<!-- Check as many boxes as possible. -->

- [x] The PR name is suitable for the release notes.
- [x] The PR follows the [contribution guidelines](https://github.com/ZcashFoundation/zebra/blob/main/CONTRIBUTING.md).
- [ ] The library crate changelogs are up to date.
- [x] The solution is tested.
- [x] The documentation is up to date.

syszery committed:
refactor(state): improve error propagation for CommitCheckpointVerifiedError

syszery committed:
Update zebra-state/src/error.rs

implements review feedback

Co-authored-by: Arya <aryasolhi@gmail.com>

syszery committed:
refactor(state): align checkpoint commit error handling with semantic commit flow

arya2 commented on zebra-state/src/error.rs:
```diff
@@ -49,6 +49,10 @@ pub enum QueueAndCommitError {
 
+    #[error("already finished committing checkpoint verified blocks: dropped duplicate block, block is already committed to the state")]
+    #[non_exhaustive]
+    AlreadyCommitted,
```
Optional: I think it would be nice to merge the `AlreadyCommitted`, `Replaced` and `ReplacedByNewer` variants into `Duplicate` and add a `source` or `location` field to `Duplicate` specifying if the duplicate was found in the queued blocks, sent blocks (`SentHashes`), non-finalized state, or finalized state. Feel free to resolve this comment if you want to leave that for a follow-up PR.

arya2 committed:
- Condenses error types and variants
- Updates `KnownBlock` request handling to check for block hashes in the state's sent hashes, fixing a bug where it's possible for part of a chain to be duplicated in the non-finalized state if a block is sent to the state via a `CommitSemanticallyVerifiedBlock` request while it's in the block write channel or before the non-finalized state channel has been updated

arya2 committed:
Updates `write_block()` to return a `CommitCheckpointVerifiedBlock` error and removes the `CloneError` variant

arya2 reviewed (commented):
This is looking great!

I started making some changes based off this branch [here](https://github.com/ZcashFoundation/zebra/tree/suggestion-commit-checkpoint-verified-errors). I'm happy to work on that a bit more or for you to takeover from there, let me know what you would prefer.

Aside from those changes, these changes will likely require some updates to `ChainSync::should_restart_sync()`.

arya2 committed:
Panic if Zebra fails to write a block to rocksdb

syszery committed:
fix: clippy lints and remove forgotten unwrap after error refactor

syszery committed:
fix Request docs and Duplicate error message

syszery committed:
fixes fmt whitespace

syszery committed:
sync: consolidate duplicate/commit checks in should_restart_sync

syszery commented on zebra-state/src/error.rs:
```diff
@@ -106,6 +106,34 @@ impl<E: std::error::Error + 'static> From<BoxError> for LayeredStateError<E> {
+
+    /// The write task exited (likely during shutdown).
+    #[error("block write task has exited. Is Zebra shutting down?")]
+    WriteTaskExited,
```
addressed by merging #9999

syszery committed:
should_restart_sync: make duplicate/commit checks more defensive

syszery commented:
Here are the TODOs we’ve identified.

* [ ] **Fix issue #2908**

* [ ] **Try to remove the `Option` from `hash_or_height`** — this requires addressing the following TODOs in the `StateService` implementation, where the `None` is currently used:
  * [ ] `queue_and_commit_to_finalized_state`: Track the latest sent height and drop any blocks below it when sending new blocks. 
  * [ ] `queue_and_commit_to_non_finalized_state`: It seems harder to blame a single block here, so the `None` case may not be meaningfully removable.

* [ ] **Beautify `CommitBlockError::Duplicate`** — this will be straightforward once the `Option` is removed.

* [ ] Minor cleanup: unify error re-exports in lib.rs.

syszery committed:
should_restart_sync: revert previous changes and propagate CommitBlockError

syszery commented:
### Update:

* Implemented `is_duplicate_request` for `CommitBlockError`.
* Updated `VerifyBlockError::Commit` to properly wrap `CommitBlockError` and introduced a `StateService` variant.
* Modified `Service::call` to propagate `CommitBlockError`.
* Reverted changes to `should_restart_sync`.

Do these changes make sense? Do we need to address also `VerifyCheckpointError`, the network errors (I understand this pertains to the `BlockDownloadVerifyError::DownloadFailed` arm), or issue #5732? Or will the `syncer` work even without a major refactor for now?

syszery committed:
sync: clean up should_restart_sync and add unit test for CommitBlockError::Duplicate

arya2 commented on zebra-state/src/request.rs:
```diff
@@ -690,6 +690,28 @@ impl MappedRequest for CommitSemanticallyVerifiedBlockRequest {
     }
 }
 
+/// Commit a checkpointed block to the state
```
Nitpick:

```suggestion
/// Commit a checkpoint-verified block to the state
```

arya2 reviewed (approved):
This looks good to merge!

syszery committed:
refactor(zebra-state): box ValidateContextError and eliminate redundant clone

syszery commented:
Hi @arya2, I’ve boxed `ValidateContextError` as we discussed. I also removed a redundant clone in `write.rs`. All tests pass locally, and there are no more clippy lints.
